### ISSUE TITLE
WebGUI is sometimes inaccessible

### ISSUE THREAD

**Issue by scienmind** [syncthing/syncthing#3401]
Occasionally after a prolonged uptime period, UI may become inaccessible.

ST log: https://gist.github.com/anonymous/d4628810f8393f7c476a4f12c5caca36
`http://localhost:8385/debug/pprof/goroutine?debug=2` contents for whatever worth: https://gist.github.com/anonymous/f31e8f71380cd6c9bc090f0d2c0b7a32

As far as UI goes, everything seems dead, though the syncing is up and running, files do fly around. 

The thing is that a while ago, following the discussion on [restartOnWakeup setting](https://forum.syncthing.net/t/question-about-restartonwakeup-setting/2222/7), I disabled it.
And this issue I'm having is appearing after wake-from-sleep. So, maybe there's a connection. 
(the original issue regarding restartOnWakeup: syncthing/syncthing#565)

Originally I suspected that it was SyncTrayzor's fault. 
The issue has been discussed here: https://github.com/canton7/SyncTrayzor/issues/292 .

Syncthing Version: v0.13.10
OS Version: Windows 10 x64

**Comment by canton7** [comment]
"UI may become inaccessible." appears to mean that it can be loaded, but returns empty (I think).

**Comment by AudriusButkevicius** [comment]
Well if you disabled that, then you are asking for it.

**Comment by scienmind** [comment]
True, but as calmh mentioned on the forum, major problems that lead to restartOnWakeup has been resolved. 
So maybe this can be fixed by a "smaller patch" rather than having full restartOnWakeup enabled constantly. Since as mentioned on the forum, restart is quite a costly procedure for battery-powered devices.

**Comment by AudriusButkevicius** [comment]
Potentially, but thats essentially just the general "no more restarts" effort, not much purpose for a separate ticket.

Also, restarts being expensive is the bigger problem, not the fact that we need to restart.

**Comment by scienmind** [comment]
OK, so for now i'll re-enable the restarts to see if the issue would indeed go away

The bottom line:
You suggest closing this for now, and keep "optimize startup efficiency" / "minimize restart necessity" as a broad todo topics in the background?

**Comment by AudriusButkevicius** [comment]
Yes.

**Comment by scienmind** [comment]
Ok, I re-enabled the "restartOnWakeup" option.
The issue is still present.

I woke the computer from sleep, after a while, network was connected, and when later I tried to access the GUI - no success: blank GUI in SyncTrayzor, or "This site can’t be reached" in external browser.
Sync itself works fine, files exchange successfully. The issue is the same as before, so the restartOnWake is probably unrelated.

In the logs i noticed the following:
`[IDIDID] 20:45:18 INFO: Restarting` - restartOnWake
`[IDIDID] 20:45:36 INFO: GUI and API listening on 10.0.0.52:GUI_PORT`
`[IDIDID] 20:45:36 INFO: Access the GUI via the following URL: https://my_computer_name:GUI_PORT/`

Now, when I unsuccessfully tried to access the GUI (around 22:2x:xx), I noticed that the computer's current IP at the time was `10.0.0.51`, and the local IP change was not logged.
Ether way, GIU is inaccessible at any of the addresses: not on `10.0.0.52:GUI_PORT` (as per ST's console), not on `10.0.0.51:GUI_PORT` (actual PC address), not on `https://my_computer_name:GUI_PORT/`, not on `localhost:GUI_PORT`

Syncthing log: https://gist.github.com/anonymous/ff20e27e9db7bb70f160ded2ef30f505

**Comment by AudriusButkevicius** [comment]
It binds to 0.0.0.0, so the IP printed does not matter.
Unless you are accessing the wrong IP (after it has changed), and expect it to work.

**Comment by canton7** [comment]
> "This site can’t be reached" in external browser.

That's not what you said before...?

> Upon trying to access the GUI via external browser, the result turned out to be the same (!). All blank.
Which leads to the following: 

(from https://github.com/canton7/SyncTrayzor/issues/292#issuecomment-232031855).

**Comment by scienmind** [comment]
> That's not what you said before...?

Sorry for misleading comment before...
External browser has reported "This site can’t be reached" in all cases.

**Comment by AudriusButkevicius** [comment]
So, are you actually expecting to access the site on the old ip?

**Comment by scienmind** [comment]
@AudriusButkevicius ofcourse not
I just tried to clarify that the gui is not accessible and to verify that some bizarre mapping has not occured

**Comment by AudriusButkevicius** [comment]
can you run equivalent of netstat on windows, which lists which processes are bound to which ports.

**Comment by canton7** [comment]
Sounds like the listener died then?

**Comment by AudriusButkevicius** [comment]
From the original stacktraces it does not look like it.

**Comment by scienmind** [comment]
Here's everything that is ST-related:
```
C:\WINDOWS\system32>netstat

Active Connections

  Proto  Local Address          Foreign Address        State
  TCP    10.0.0.51:22000        ANOTHER_PC_1:59248  ESTABLISHED
  TCP    10.0.0.51:22000        ANOTHER_PC_2:58439        ESTABLISHED
  TCP    10.0.0.51:49801        discovery-v4-2:https   TIME_WAIT
  TCP    10.0.0.51:49802        discovery-v4-3:https   TIME_WAIT
  TCP    [IPV6]:22000  [IPV6]:61376  ESTABLISHED
  TCP    [IPV6]:61376  [IPV6]:22000  ESTABLISHED
```

**Comment by AudriusButkevicius** [comment]
Right, can you get the stack traces in that case?

**Comment by scienmind** [comment]
You mean sttrace? If so, which flags to enable?

**Comment by AudriusButkevicius** [comment]
no the  profile via pprof url

**Comment by scienmind** [comment]
`STPROFILER/debug/pprof/goroutine?debug=2`: 
https://gist.github.com/anonymous/8f1dd814ae9bfe0d652ce57d7118d919

**Comment by AudriusButkevicius** [comment]
Nope, the listener is alive and well:
https://gist.github.com/anonymous/8f1dd814ae9bfe0d652ce57d7118d919#file-gistfile1-txt-L1073

**Comment by canton7** [comment]
Hmm, actually, the fact that SyncTrayzor hasn't complained suggests that the rest api might be alive and well...

**Comment by calmh** [comment]
> [IDIDID] 20:45:36 INFO: GUI and API listening on 10.0.0.52:GUI_PORT

This looks like it's *configured* to bind to 10.0.0.52 by configuration, which will of course break if the computer doesn't actually have that address.

**Comment by scienmind** [comment]
btw, the common pattern for all occurrences I had so far:

1. Put PC to sleep at location A with WiFi AA 
2. Travel to location B with WiFis BB, CC, DD (all are accessible by the PC)
3. Wake the PC at location B

**Comment by calmh** [comment]
No really, if it says "listening on 10.0.0.52:..." as opposed to listening on 0.0.0.0:... or 127.0.0.1:..., that's because it's been told to bind specifically to that address. It'll break if the addressing changes. Why is it bound to that address?

**Comment by scienmind** [comment]
> Why is it bound to that address?

Thats the question...

In both syncthing's and synctrayzor's settings, GUI listen address is configured to `my_computer_name:GUI_PORT` , which supposed to be fine...

**Comment by wweich** [comment]
> configured to `my_computer_name:gui_port`

That's the  problem. Why did you set it like that? Syncthings UI should either be only accessible locally (`127.0.0.1:GUI_PORT`) or also from other machines (`0.0.0.0:GUI_PORT`). Setting it to the computername or external IP makes no sense.

A service cannot be bound to a name, but only to an IP. If your computer name results in 10.0.0.52 when syncthing is started, it will stay bound to that IP until it is restarted.

**Comment by scienmind** [comment]
Hmm, I usually set it up this way to access PC by `pc_name: ...`
And I have been doing the same with ST for quite a while now.
I guess the recent addition of multiple "accessible" networks per-location made this issue to surface... 

I tested it, and indeed, local device IP switch is the thing causing "inaccessible" GUI problem, easily reproducible.
Now, I changed all ST configs to `0.0.0.0:port`, retested and all fixed!!!


Console output changed from:
```
[IDIDID] 20:45:36 INFO: GUI and API listening on 10.0.0.5x:GUI_PORT
[IDIDID] 20:45:36 INFO: Access the GUI via the following URL: https://my_computer_name:GUI_PORT/
```
To:
```
[IDIDID] 13:52:38 INFO: GUI and API listening on [::]:PORT
[IDIDID] 13:52:38 INFO: Access the GUI via the following URL: https://127.0.0.1:PORT/
```
And I still can access it from external devices by `pc_name: ...`

Thanks @wweich for pointing it out!
Issue closed indeed!

(and I'll probably switch "restartOnWake" off again to keep on testing it, since apparently there are no real issues with it so far)
